File: stock/src/infra/rabbitmq/listeners/ProductCreatedListener.ts
import { Listener } from './Listener'
import { Message } from 'amqplib'
import { Exchanges } from '../exchanges'
import { ProductCreatedEvent } from '../events/ProductCreatedEvent'
import { Product } from '../../../models/product'

export class ProductCreatedListener extends Listener<ProductCreatedEvent> {
	readonly exchange = Exchanges.ProductCreated

	constructor(queueName: string) {
		super(queueName)
	}

	async consume(content: ProductCreatedEvent, msg: Message): Promise<void> {
		const { id, title, description, price } = content

		console.log(content.id)

		const product = Product.build({
			id,
			title,
			description,
			price,
		})

		await product.save()

		console.log(`Product created: ${product._id}`)
	}
}

File: buyer/src/main/java/com/madness/microservice/infra/rabbitmq/listeners/ProductCreatedListener.java
package com.madness.microservice.infra.rabbitmq.listeners;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.madness.microservice.infra.mongo.MongoDbConnection;
import com.madness.microservice.infra.rabbitmq.Exchanges;
import com.madness.microservice.infra.rabbitmq.events.ProductCreatedEvent;
import com.madness.microservice.models.Product;
import com.mongodb.client.MongoCollection;

@Singleton
public class ProductCreatedListener extends Listener<ProductCreatedEvent> {

  @Override
  protected Exchanges exchange() {
    return Exchanges.ProductCreated;
  }

  @Override
  protected String queue() {
    return "buyer:product:created";
  }

  @Inject
  public MongoDbConnection conn;

  @Override
  public void consume(final ProductCreatedEvent data) {

    System.out.println(data.id);

    var collection = conn.collection("products", Product.class);

    var product = new Product(data.id);

    System.out.println("Inserting new product");

    collection.insertOne(product);

    System.out.println("Product inserted.");
  }

}